### PR TITLE
`unnormalize()` with grouped data

### LINKED ISSUE #375: `unstandardize()` doesn't work with grouped data
No problem with ungrouped data:
``` r
library(datawizard)

foo <- iris |> 
  # data_group(Species) |> 
  standardize(select = Sepal.Length) |> 
  unstandardise(select = "Sepal.Length")

waldo::compare(foo, iris, ignore_attr = TRUE)
#> ✔ No differences
```
Doesn't work with grouped data:
``` r
library(datawizard)

iris |> 
  data_group(Species) |>
  standardize(select = Sepal.Length) |> 
  unstandardise(select = "Sepal.Length")
#> Error: Cannot (yet) unstandardize a `grouped_df`.
```

### PR DESCRIPTION
**Pull request by etiennebacher**
Close #375. 

``` r
library(datawizard)

x <- iris |> 
  data_group(Species) |>
  normalize(select = Sepal.Length) |> 
  unnormalize(select = Sepal.Length) |> 
  data_ungroup()

identical(x, iris)
#> [1] TRUE

x <- iris |> 
  data_group(Species) |>
  standardize(select = Sepal.Length) |> 
  unstandardize(select = Sepal.Length) |> 
  data_ungroup()

identical(x, iris)
#> [1] TRUE
```

The idea is to store the `dw_transformer` attributes along the `groups` attributes so that we can then recover the `dw_transformer` attributes for each group separately. We then need to pass the `dw_transformer` attributes to `normalize.data.frame()` and then to `normalize.numeric()`. 

Same thing for `unstandardize()`.

Note: there are probably ways to improve this in terms of performance

### CODE REVIEW

**Comment by etiennebacher** [comment]
@strengejacke if you're ok with this code then I can expand it to `standardize()`, `center()`, etc.

**Comment by strengejacke** [comment]
I'm not sure, but I think @mattansb started with one of the `un*()` functions? Maybe you can better review this PR?

File: R/unstandardize.R [hunk 1]
```diff
@@ -143,8 +143,69 @@ unstandardize.grouped_df <- function(x,
...
+  # works only for dplyr >= 0.8.0
+  grps <- attr(x, "groups", exact = TRUE)[[".rows"]]
+
+  x <- as.data.frame(x)
+
+  for (i in select) {
+    if (is.null(info$groups[[paste0("attr_", i)]])) {
+      insight::format_error(
+        paste("Couldn't retrieve the necessary information to unnormalize",
+              text_concatenate(i, enclose = "`"))
```
Comment (strengejacke) [inline R/unstandardize.R:174]: Isn't this `unstandardize()`?

**Comment by mattansb** [comment]
Daniel keeps thinking that @DominiqueMakowski 's code is mine 😅

**Comment by etiennebacher** [comment]
The code is incomplete for `unstandardize()`, it doesn't work with the example in #375

**Comment by strengejacke** [comment]
> Daniel keeps thinking that @DominiqueMakowski 's code is mine 😅

But see _effectsize_ news, version 0.5.0 `unstandardize()` was moved to _datawizard_, and I think you made one of the first implementations of those `un*()` functions (I think I then tried to adopt this for `unnormalize()`).

**Comment by mattansb** [comment]
I'm like 90% certain it was Dom, and then I maintained the code.

**Comment by etiennebacher** [comment]
I'll try to refactor later but it seems to work now:
``` r
library(datawizard)

x <- iris |> 
  data_group(Species) |>
  standardize(select = Sepal.Length) |> 
  unstandardise(select = "Sepal.Length") |> 
  data_ungroup()

identical(x, iris)
#> [1] TRUE
```

**Comment by etiennebacher** [comment]
@strengejacke @DominiqueMakowski or @mattansb I don't know who's the best placed to review but can one of you do it? 😄

File: R/unnormalize.R [hunk 2]
```diff
@@ -17,10 +17,28 @@ unnormalize.default <- function(x, ...) {
...
-  to_range <- attr(x, "to_range")
+
+  # if function called from the "grouped_df" method, we use the dw_transformer
+  # attributes that were recovered in the "grouped_df" method
+
+  dots <- match.call(expand.dots = FALSE)[["..."]]
+  grp_attr_dw <- eval(dots$grp_attr_dw, envir = parent.frame(1L))
+
+  if (!is.null(grp_attr_dw)) {
+    names(grp_attr_dw) <- gsub(".*\\.", "", names(grp_attr_dw))
```
Comment (strengejacke) [inline R/unnormalize.R:28]: I can't say if this regex pattern is correct, but I guess it is and it's covered by tests?

**Review by strengejacke — APPROVED** [review]

File: R/unnormalize.R (hunk 2 above)
Comment (etiennebacher) [reply R/unnormalize.R:28]: Thanks, I'll add some extra checks with other variables

**Comment by etiennebacher** [comment]
Last thing, I think we should harmonize the behavior of `unstandardize()` and `unnormalize()` when they don't have the necessary info. Currently, `unstandardize()` fails but `unnormalize()` only gives a warning:

``` r
library(datawizard)

unstandardize(mtcars, "mpg")
#> Error: You must provide the arguments `center`, `scale` or `reference`.

unnormalize(head(mtcars), "mpg")
#> Warning: Can't unnormalize variable. Information about range and/or minimum value
#>   is missing.
#>                    mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
#> Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
#> Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
#> Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
#> Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1
```
IMO both should fail. @strengejacke do you agree?

**Comment by strengejacke** [comment]
yes.

**Pull request merged**